src/main/java/com/attacktimer/AttackTimerMetronomePlugin.java

```diff
@@ -174,42 +281,129 @@ private int adjustSpeedForLeaguesIfApplicable(int baseSpeed)
         return baseSpeed;
     }
 
-    private int getWeaponSpeed()
+    private void setAttackDelay()
     {
-        boolean isPlayerCasting = isPlayerCasting();
-        int itemId = getWeaponId();
-        // We cannot use the weapon type to determine this as the shadow is also
-        // a powered stave which has a 5-tick attack. Though weapon stats might
-        // be better.
-        boolean isFourTick = WeaponIds.FourTickStaves().contains(itemId);
-        if (isPlayerCasting && !isFourTick)
+        // If the weapon Id isn't consistent across this method we race with the player if they're switching
+        // mid tick, so we have to cache this here.
+        int weaponId = getWeaponId();
+        AnimationData curAnimation = AnimationData.fromId(client.getLocalPlayer().getAnimation());
+        PoweredStaves stave = PoweredStaves.getPoweredStaves(weaponId, curAnimation);
+        boolean matchesSpellbook = matchesSpellbook(curAnimation);
+        // Slow path - first blindly guess the speed without checking projectile
+        attackDelayHoldoffTicks = getWeaponSpeed(weaponId, null, stave, curAnimation, matchesSpellbook);
+        if (stave == null)
         {
-            // Unfortunately, the trident and toxic trident share animations with wave-tier spells.
-            //
-            // Assume that if the user is not wielding a trident and isCasting, they are autocasting
-            // or manual casting from a spellbook.
+            // Fast path, no need to check the stave code
+            return;
+        }
+
+        // Now do more thorough checks (most weapons don't need this stuff and exit early, melee and ranged) -
+        // However if it's a powered staff and a casting animation tied to one of the staves. Then we don't
+        // actually know if the staff is casting a builtin spell or something off the current spellbook. We
+        // can do slightly better by checking the projectile we fired, as the powered staves have custom
+        // projectiles. But we start in the above getWeaponSpeed call to just assume holding the staff means
+        // 4-tick.
+        Integer firedProjectile = getPlayerFiredProjectile();
```

Comment:
I would prefer to not depended on a deprecated API. Can we remove the projectile tracking code?

Reply:
Affirmative, I will push a commit removing it 🫡